### PR TITLE
Java: Add synthetic fields; model Commons Lang's MutableObject type

### PR DESCRIPTION
**Pull request by smowton**
Based on https://github.com/github/codeql/pull/5772

This adds support for synthetic fields (for now, specifically in external flow specifications), then uses one to model Commons Lang's `MutableObject` type.

### CODE REVIEW

**Comment by Marcono1234** [comment]
Not trying to discredit this pull request, it sounds extremely useful for cases where multiple synthetic fields exist, but for `MutableObject` wouldn't the existing dataflow capabilities suffice? Or is this partially done here for demonstration purposes?

**Comment by smowton** [comment]
It actually does add value even with just one field: it enables us to use value flow instead of taint flow. Value flow has to be type-preserving, so we need an Object-typed field, as opposed to the MutableObject or Mutable<Object>-typed container. With value flow it's possible to preserve access paths: for example, with taint rules, this example doesn't work:

```
MyClass c = new MyClass();
c.field = unsafe(); // Access path created; we expect a `.field` get to re-expose unsafe data
Mutable<MyClass> container = new Mutable<MyClass>(c);
sink(container.get().field);
```

It doesn't work because taint rules only apply with an empty access path: we can say that if `c` is tainted then `container` is tainted then `container.get()` is tainted, but not that latent taint that requires a field access to re-expose is preserved. Value-flow on the other hand can do that: if we document that the *exact value* of `c` is stored in some field of `container` that is fetched by its `get()` method, then the fetch from `.field` will work as expected.

File: java/ql/src/semmle/code/java/dataflow/ExternalFlow.qll [hunk 1]
```diff
@@ -834,3 +867,30 @@ predicate sinkNode(Node node, string kind) {
...
 }
+
+/**
+ * A synthetic field referred to in function models. Extend this type to introduce a new
+ * synthetic field.
+ *
+ * The owning type has no significance except to provide a source location, and to
+ * aid in avoiding name clashes.
+ */
+abstract class SyntheticField extends Unit {
```
[the remark this answers is not in the document]
Comment (aschackmull) [reply java/ql/src/semmle/code/java/dataflow/ExternalFlow.qll:878]: This observation is spot on. Declaring `class SyntheticField extends Unit` means that there's only ever one `SyntheticField`.

[the remark this answers is not in the document]
Comment (smowton) [reply java/ql/src/semmle/code/java/dataflow/ExternalFlow.qll:878]: This was written back when I thought we identified instances with rows in the charpred table (i.e., with owningType/fieldName tuples), will change...

[the remark this answers is not in the document]
Comment (smowton) [reply java/ql/src/semmle/code/java/dataflow/ExternalFlow.qll]: Rewritten in the style of the CSV row extension point, and rebased. Ready for re-review.

**Comment by smowton** [comment]
^^ Checked, these diffs are reversed. Actually we gained 3 rows, as expected.

**Comment by aschackmull** [comment]
@tamasvajk The action is a bit spammy - is there any way to easily do something about that? E.g. don't post another comment if it already posted an identical comment?

**Comment by tamasvajk** [comment]
> @tamasvajk The action is a bit spammy - is there any way to easily do something about that? E.g. don't post another comment if it already posted an identical comment?

Hmm, good point. I'll figure something out. (I can probably figure out what was the `run_id` of a previous execution, and then compare the artifacts before adding a new comment.)

**Comment by smowton** [comment]
@aschackmull this is ready for review

**Comment by aschackmull** [comment]
Synthetic field support superseded by https://github.com/github/codeql/pull/6291

**Comment by smowton** [comment]
Rebased. I've kept the improvement to `clearsContent`, which implicitly assumes these synthetic fields are more like ordinary fields than like `ArrayElement` etc which cannot be cleared. If we want Synthetic things which aren't subject to that sort of strong update then we'll need either a way to customise them or a cousin to SyntheticField that doesn't appear in `clearsContent`.

**Comment by joefarebrother** [comment]
> Rebased. I've kept the improvement to `clearsContent`, which implicitly assumes these synthetic fields are more like ordinary fields than like `ArrayElement` etc which cannot be cleared. If we want Synthetic things which aren't subject to that sort of strong update then we'll need either a way to customise them or a cousin to SyntheticField that doesn't appear in `clearsContent`.

In fact I have a use case for this kind of thing; in https://github.com/github/codeql/pull/6174 I have a Table type in which its Rows and Columns are like MapKey so shouldn't be cleared. So far I've been working around it by using things like `SyntheticField[Table.row] of MapKey of ...`, which is less satisfactory.

**Comment by aschackmull** [comment]
I don't think we should try to include clearing and it also won't work as-is. The reason it won't work is that clearing flow isn't summarised, so if it happens inside a method then callers won't see it happening. And the reason for not summarising clearing flow is that all the calculated flow by the library is existential flow ("there exists at least one path such that"), for which something like setter- and getter-flows can be summarised and incorporated at the call site, but clearing flow is essentially a barrier rather than flow, so it would have to be expressed in terms of universal flow ("for all paths it holds that ..." or "no path exists such that ...").

**Comment by aschackmull** [comment]
Clearing-as-summary is partially supported by `FlowSummaryImpl.qll`, but it isn't something that's currently hooked up to csv rows, as it would need special handling there (it's a separate kind of summary).

**Comment by aschackmull** [comment]
Hmm, I see the test does seem to indicate that it actually works... I wonder why that is. I'm not sure it should.

**Comment by smowton** [comment]
@aschackmull dropped clearing for synthetic fields. @joefarebrother this means synthetic fields always behave like `Element` et al so your case can probably be simplified.

**Review by aschackmull — APPROVED** [review]

**Pull request merged**